ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsFetch.scala

```diff
@@ -295,70 +295,79 @@ private class ContractsFetch(
-        val (idses, lastOff) = (startOffset, disableAcs) match {
-          case (LedgerBegin, false) =>
-            val stepsAndOffset = builder add acsFollowingAndBoundary(txnK)
-            stepsAndOffset.in <~ getActiveContracts(
-              jwt,
-              ledgerId,
-              transactionFilter(parties, List(templateId)),
-              true,
-            )(lc)
-            (stepsAndOffset.out0, stepsAndOffset.out1)
-
-          case (AbsoluteBookmark(_), _) | (LedgerBegin, true) =>
-            val stepsAndOffset = builder add transactionsFollowingBoundary(txnK)
-            stepsAndOffset.in <~ Source.single(startOffset)
-            (
-              (stepsAndOffset: FanOutShape2[_, ContractStreamStep.LAV1, _]).out0,
-              stepsAndOffset.out1,
-            )
-        }
+    // skip if *we don't use the acs* (which can read past absEnd) and current
+    // DB is already caught up to absEnd
+    if (startOffset == AbsoluteBookmark(absEnd.toDomain))
+      fconn.pure(startOffset)
+    else {
+      val graph = RunnableGraph.fromGraph(
+        GraphDSL.createGraph(
+          Sink.queue[ConnectionIO[Unit]](),
+          Sink.last[BeginBookmark[domain.Offset]],
+        )(Keep.both) { implicit builder => (acsSink, offsetSink) =>
+          import GraphDSL.Implicits._
+
+          val txnK = getCreatesAndArchivesSince(
+            jwt,
+            ledgerId,
+            transactionFilter(parties, List(templateId)),
+            _: lav1.ledger_offset.LedgerOffset,
+            absEnd,
+          )(lc)
+
+          // include ACS iff starting at LedgerBegin
+          val (idses, lastOff) = (startOffset, disableAcs) match {
+            case (LedgerBegin, false) =>
+              val stepsAndOffset = builder add acsFollowingAndBoundary(txnK)
+              stepsAndOffset.in <~ getActiveContracts(
+                jwt,
+                ledgerId,
+                transactionFilter(parties, List(templateId)),
+                true,
+              )(lc)
+              (stepsAndOffset.out0, stepsAndOffset.out1)
+
+            case (AbsoluteBookmark(_), _) | (LedgerBegin, true) =>
+              val stepsAndOffset = builder add transactionsFollowingBoundary(txnK)
+              stepsAndOffset.in <~ Source.single(startOffset)
+              (
+                (stepsAndOffset: FanOutShape2[_, ContractStreamStep.LAV1, _]).out0,
+                stepsAndOffset.out1,
+              )
+          }
 
-        val transactInsertsDeletes = Flow
-          .fromFunction(
-            jsonifyInsertDeleteStep((_: InsertDeleteStep[Any, lav1.event.CreatedEvent]), templateId)
-          )
-          .via(conflation)
-          .map(insertAndDelete)
+          val transactInsertsDeletes = Flow
+            .fromFunction(
+              jsonifyInsertDeleteStep(
+                (_: InsertDeleteStep[Any, lav1.event.CreatedEvent]),
+                templateId,
+              )
+            )
+            .via(conflation)
```

Comment:
https://github.com/digital-asset/daml/pull/16401/files#diff-73c6cf12a0baa5992ba41a5b30d92e9ffbe234debd5876949b5cff238de312e3R337 is needed as well. 457af4b780d06871d1a6ba685bd48733e5491dd3